From 9c2ae13c2fd1e9d351178f9f21f7f030fd779e33 Mon Sep 17 00:00:00 2001
From: Andi Kleen <andi@firstfloor.org>
Date: Sun, 28 Mar 2021 10:28:16 -0700
Subject: [PATCH 06/39] static_call: Make static call functions visible

Since static call functions are referenced from inline assembler
they must be __visible and global with gcc LTO. Fix this for the static
call self tests.

Cc: peterz@infradead.org
Cc: rostedt@goodmis.org
Signed-off-by: Andi Kleen <andi@firstfloor.org>
---
 kernel/static_call.c | 8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

diff --git a/kernel/static_call.c b/kernel/static_call.c
index 2c5950b0b90e..6d332c0c9134 100644
--- a/kernel/static_call.c
+++ b/kernel/static_call.c
@@ -503,7 +503,9 @@ long __static_call_return0(void)
 
 #ifdef CONFIG_STATIC_CALL_SELFTEST
 
-static int func_a(int x)
+int __visible sc_func_a(int x); /* Avoid warning */
+
+int __visible sc_func_a(int x)
 {
 	return x+1;
 }
@@ -513,7 +515,7 @@ static int func_b(int x)
 	return x+2;
 }
 
-DEFINE_STATIC_CALL(sc_selftest, func_a);
+DEFINE_STATIC_CALL(sc_selftest, sc_func_a);
 
 static struct static_call_data {
       int (*func)(int);
@@ -522,7 +524,7 @@ static struct static_call_data {
 } static_call_data [] __initdata = {
       { NULL,   2, 3 },
       { func_b, 2, 4 },
-      { func_a, 2, 3 }
+      { sc_func_a, 2, 3 }
 };
 
 static int __init test_static_call_init(void)
-- 
2.30.2

